source/slang/slang-ir-byte-address-legalize.cpp

```diff
@@ -324,9 +322,19 @@ struct ByteAddressBufferLegalizationContext
             // legalization if the array type isn't in the right form
             // for us to proceed.
             //
+
             if (auto elementCountInst = as<IRIntLit>(arrayType->getElementCount()))
             {
-                return emitLegalSequenceLoad(type, buffer, baseOffset, immediateOffset, kIROp_MakeArray, arrayType->getElementType(), elementCountInst->getValue());
+                // Emit an aligned load operation on an array when using a LoadAligned inst.
+                // Else, fallback to scalarizing the loads.
+                if (!isValid(aligned))
```

Comment:
I am not sure this is correct.

If I have:
```
struct MyType
{
    float4 arr[4];
};

buffer.LoadAligned<MyType>(offset = 16, alignment = 16)
```
The current logic will cause this to be translated into a load from `[0]`, right?

In general, we need to have:
```
ByteAddressBuffer::Load<T>(uint offset);
ByteAddressBuffer::Load<T>(uint offset, uint alignment);

ByteAddressBuffer::LoadAligned<T>(uint offset)
{
    return Load<T>(offset, __alignOf<T>());
}

```

We almost always want to load the data from an equivalent `StructuredBuffer<uint4>` or `StructuredBuffer<uint2>` if we know `offset + immediateOffset` is aligned to 16 or 8 bytes. The `alignment` parameter if provided by the user, tells us what `offset` is aligned to if it is not a known compile time value.

Reply:
Note that in this case, sizeof(MyType) is 64, but `Load<MyType>(offset =16,  alignment=16)` should still result in aligned vectorized reads, despite `offset=16` is not aligned to the boudary of `MyType`'s size (64).

Reply:
I think I understand now. Instead of having align as a true / false value, we can have it as an integer value that the user can pass, and use that information to get the exact alignment check.
So, within the stdlib, all the references of Load/ LoadAligned would be modified to something like below.
```
    [__NoSideEffect]
    [ForceInline]
    [require(cpp_cuda_glsl_hlsl_spirv, byteaddressbuffer_rw)]
    T Load<T>(int location)
    {
        return __byteAddressBufferLoad<T>(this, location, 0);
    }

    [__NoSideEffect]
    [ForceInline]
    [require(cpp_cuda_glsl_hlsl_spirv, byteaddressbuffer_rw)]
    T Load<T>(int location, int align)
    {
        return __byteAddressBufferLoad<T>(this, location, align);
    }

    [__NoSideEffect]
    [ForceInline]
    [require(cpp_cuda_glsl_hlsl_spirv, byteaddressbuffer_rw)]
    T LoadAligned<T>(int location)
    {
        return Load<T>(location, __alignOf<T>());
    }
```
Does this look ok?, we may want to default align = -1, because, align = 0, now also means that the resource is aligned.

Reply:
I think `alignment = 1` as default makes sense because 1 byte is the minmum alignment you can get (which means no alignments).